### PR TITLE
Move grid information off layer

### LINKED ISSUE #1690: Move grid information off individual layers
## Refactor

As part of improvements to our handling of transforms I propose we move `grid` mode off of the individual layers and have all of that done at the level of the viewer

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR would move the grid information off the layer and close #1690, as layers shouldn't know where they are in the grid. 

I'd like to address #860 before merge as best we can too. I'll be updating this as we go, this is currently still a work in progress.

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by sofroniewn** [comment]
This is still a work in progress. 3D rendering and multiscale grid need to be fixed

This will though now propagate cursor position back to the layer appropriately.

I don't love what's happening here though, the grid does break what previously were some nice abstractions with `world` coordinates, because now we have `world` and `worldgrid`, but it's sort of unavoidable. Maybe we can make this a little nicer before merge

**Comment by sofroniewn** [comment]
Ok this is now ready for review @jni, I can't say I love love it, as I have fears that I'm trading complexity in one part of the code base for another, particularly around the coordinate handling and offsetting. I do think though that removing `layer.translate_grid` is unambiguously a good thing as we try and simplify our layer models though, so I take that at least as a positive! Any ideas on how to make things simpler would be much appreciated.

After this merges I'd like to fix #1836 and then I think we're ready to continue looking at #1775 and improving the slicing. All in all though I'm pretty happy with our new camera, cursor, and grid objects

File: napari/_qt/qt_viewer.py
Comment (jni) [inline napari/_qt/qt_viewer.py:726]: Sometimes I really hate `black`...

**Review by jni — CHANGES_REQUESTED** [review]
@sofroniewn I think what's missing here is a Canvas model, and honestly I don't think this is a good intermediate step to get there. If we want another frontend to this, everything will have to move right back from VisPy and Qt to model files...

What do you think? Happy to discuss on Zulip tomorrow morning my time.

**Comment by sofroniewn** [comment]
Yeah, makes sense, I had my doubts about this too. I might pull out the grid.shape rename as I think it’s better than grid.size, but otherwise this can wait

**Comment by sofroniewn** [comment]
> I might pull out the grid.shape rename as I think it’s better than grid.size, but otherwise this can wait

Ok, I've done this now in #1847, I'm turning this to draft until we get a canvas object or have more ideas

File: napari/layers/base/base.py
Comment (jni) [inline napari/layers/base/base.py:467]:
```suggestion
                "The layer.translate_grid attribute is deprecated and will be removed after version 0.4.3."
                " Layers no longer know about their position in a grid and this property should no longer"
                " be used. It now returns a list of 0."
```

File: napari/components/grid.py
Comment (jni) [inline napari/components/grid.py:21]: @sofroniewn this docstring needs an explanation of `.size`, which is still an attribute of the class.

File: napari/components/viewer_model.py
Comment (jni) [inline napari/components/viewer_model.py:444]: I kind of think that this logic should use a transform (for which we have models) rather than this custom translation. Would that be easy enough to add?

File: napari/_vispy/vispy_base_layer.py
Comment (jni) [inline napari/_vispy/vispy_base_layer.py:99]: @sofroniewn to me it's still pretty concerning that the grid transform is not in the layer but it *is* in the VisPy layer...

File: napari/_tests/test_with_screenshot.py
Comment (jni) [inline napari/_tests/test_with_screenshot.py:319]: What is `actual_shape` compared to `shape`? This API might benefit from a rethink...

Comment (jni) [reply napari/_tests/test_with_screenshot.py:319]: and `vis.translate_grid[:2][::-1]` is super mysterious here...

**Review by jni — APPROVED** [review]
@sofroniewn I'll approve since you mentioned this is blocking, but I still have some real concerns about where this is heading... It's weird that things are on a VisPy layer but not a layer itself...

Other than that, I've left minor comments that should be easy to fix!

**Comment by sofroniewn** [comment]
> @sofroniewn I'll approve since you mentioned this is blocking, but I still have some real concerns about where this is heading... It's weird that things are on a VisPy layer but not a layer itself...

Opps @jni - I requested you to review the wrong PR, I wanted you to look at #1847 which is much more minimal, just changes `size` to `shape`. I share reservations about this one too and want to spend more time thinking about it too see https://github.com/napari/napari/pull/1827#issuecomment-722875601, though I do think the overall goal still makes sense

File: napari/_qt/qt_viewer.py
Comment (Czaki) [reply napari/_qt/qt_viewer.py:726]: negate `if` clause and call `continue`. This will reduce indentation

File: napari/_qt/qt_viewer.py
Comment (Czaki) [inline napari/_qt/qt_viewer.py:712]:
```suggestion
        for i, layer in enumerate(reversed(self.viewer.layers)):
```

**Comment by sofroniewn** [comment]
This is still on hold, we'll have to see how some other refactoring goes, but big picture I do still like this

**Comment by sofroniewn** [comment]
We're going to hold on this until we have a better multicanvas model, so I will close this now but leave the underlying issue open

**Pull request closed without merging**
